test/object-store/sync/flx_sync.cpp

```diff
@@ -3299,6 +3300,187 @@ TEST_CASE("flx: bootstrap changesets are applied continuously", "[sync][flx][app
     CHECK(user_commit_version == bootstrap_version + 1);
 }
 
+TEST_CASE("flx: async open + register subscription callack while bootstrapping", "[sync][flx][async-open]") {
+    std::condition_variable cv;
+    std::mutex mutex;
+    std::atomic<bool> async_open_invoked = false;
+    std::atomic<bool> subscription_invoked = false;
+    std::atomic<bool> async_open_completed = false;
+    FLXSyncTestHarness harness("flx_bootstrap_batching");
+    auto foo_obj_id = ObjectId::gen();
+    harness.load_initial_data([&](SharedRealm realm) {
+        CppContext c(realm);
+        Object::create(c, realm, "TopLevel",
+                       std::any(AnyDict{{"_id", foo_obj_id},
+                                        {"queryable_str_field", "foo"s},
+                                        {"queryable_int_field", static_cast<int64_t>(5)},
+                                        {"non_queryable_field", "created as initial data seed"s}}));
+    });
+    SyncTestFile config(harness.app()->current_user(), harness.schema(), SyncConfig::FLXSyncEnabled{});
+
+    auto subscription_callback = [&](std::shared_ptr<Realm> realm) {
+        REQUIRE(realm);
+        auto table = realm->read_group().get_table("class_TopLevel");
+        Query query(table);
+        auto subscription = realm->get_latest_subscription_set();
+        auto mutable_subscription = subscription.make_mutable_copy();
+        mutable_subscription.insert_or_assign(query);
+        subscription_invoked = true;
+        mutable_subscription.commit();
+    };
+    auto async_open_complete = [&](ThreadSafeReference ref, std::exception_ptr err) {
+        if (err) {
+            // there has been an error, the test must fail.
+            async_open_invoked = true;
+            cv.notify_one();
+            return;
+        }
+        auto realm = Realm::get_shared_realm(std::move(ref));
+        REQUIRE(realm);
+        async_open_invoked = true;
+        async_open_completed = true;
+        auto table_ref = realm->read_group().get_table("class_TopLevel");
+        REQUIRE(table_ref);
+        REQUIRE(table_ref->get_column_count() == 4);
+        REQUIRE(table_ref->get_column_key("_id"));
+        REQUIRE(table_ref->get_column_key("queryable_str_field"));
+        REQUIRE(table_ref->get_column_key("queryable_int_field"));
+        REQUIRE(table_ref->get_column_key("non_queryable_field"));
+        cv.notify_one();
+    };
+
+    config.sync_config->subscription_initializer = subscription_callback;
+    auto async_open = Realm::get_synchronized_realm(config);
+    async_open->start(async_open_complete);
+
+    std::unique_lock<std::mutex> lock(mutex);
+    cv.wait(lock, [&] {
+        return async_open_invoked.load();
+    });
+    REQUIRE(subscription_invoked.load());
+    REQUIRE(async_open_invoked.load());
+    REQUIRE(async_open_completed.load());
+
+    SECTION("subscription not run because realm alredy opened no error") {
```

Comment:
why is this the case? where is realm opened?

Reply:
```
auto async_open = Realm::get_synchronized_realm(config);
async_open->start(async_open_complete);  //opened here

std::unique_lock<std::mutex> lock(mutex);
cv.wait(lock, [&] {
        return async_open_invoked.load();
});
REQUIRE(subscription_invoked.load());
REQUIRE(async_open_invoked.load());
```
And then it gets reopened in the SECTION
```
 SECTION("subscription not run because realm alredy opened no error") {
        subscription_invoked = async_open_invoked = async_open_completed = false;
        auto async_open = Realm::get_synchronized_realm(config);
        async_open->start(async_open_complete);
        
        .....
```